### DISCUSSION TITLE
Create live photo from video and photo

### DISCUSSION THREAD

**Discussion by Yorian** [RhetTbull/osxphotos#1398] [General]
Hi,

I have some old live photos which used to be on Android. Would it be possible to create "live photos" in the Photos app using osxphotos?

**Comment by oPromessa** [comment]
Hello @Yorian   

Roughly... Live Photos on iOS are a pair/combination of an .HEIC image file and corresponding .MOV video file. As far as I know the .HEIC file has a tag which indicates there should be a corresponding video file and, such file existing, apps and Photos can then understand them as a Live Photo.  

They can be loaded directly to Photos and osxphotos can then export the pair (with appropriate options!). 

Osxphotos can also import into Photos pics and photos and I trust also Live Photos -- did not test it though ;). 

Now to your question: When you say Live Photos from Android don't know if Android follows the same structure and approach as iOS iPhones and if Photos will recognize it. 
- Have you tried loading the Android Live Photos directly into Photos? Does photos recognizes them as Live Photos?

**Reply by Yorian** [reply]
Thanks for the quick reply! I did try loading them but Photos does not recognize them as Live Photos. This is likely due to the the tags that are missing in the photos/videos.

**Reply by RhetTbull** [reply]
@Yorian, @oPromessa is right that there's a tag that needs to be in the image and video to be associated as a Live Photo. I've not done extensive testing but it appears to require the "Content Identifier" tag which is `MakerNotes:ContentIndentifer` in the HEIC or JPG and `QuickTime:ContentIdentifier` in the video file. The value is a universally unique identifier (UUID) that must match for Photos to correlate the two. The UUID looks like: `4636F170-F379-42AC-9CEA-20233E95936F` (but each value is unique to the photo/video pair). You can set this via [exiftool](https://exiftool.org/) using `exiftool -ContentIdentifier="4636F170-F379-42AC-9CEA-20233E95936F" IMG_1234.jpg IMG_1234.mov`  I've tested this with both HEIC and JPG + MOV and when you drag the two images into Photos they are matched as a live photo pair. `osxphotos import` can also be used to import them. 

You can produce a UUID using this python one-liner: `python -c "import uuid; print(str(uuid.uuid4()).upper())"` which will print out a unique UUID in the correct format. If you have any scripting skills, you could write a script to find matching JPG+MOV files, generate the UUID, and apply the tag using exiftool.

It might be a useful feature for `osxphotos import` to add ability to do this automatically. Perhaps `osxphotos import --auto-live` and let osxphotos take care of tagging the images.

**Reply by RhetTbull** [reply]
I've opened #1399 to add this to `osxphotos import`.

**Reply by RhetTbull** [reply]
@yorian I'm happy to take a crack at writing a script to do this. If you have a pair of JPG+mov images you don't mind sharing for testing, send to me at osxphotos.py@gmail.com

**Reply by Yorian** [reply]
Hi,

I think you are correct for as far as I had read that both the image and the video need a matching UUID (they do not have this currently). Setting this with exiftool doesn't seem to work however (tried: `exiftool -ContentIdentifier="1a20cf6b-7b56-433b-b60f-6781966cf7c0" 20200809_161134.jpeg` but that returns that it has not changed anything). I think this doesn't work because, from what I've read, you can not create new makernotes (or tags in the makernotes), updating the ContentIdentifier might work as long as both already have this property I think. Obviously I'd love to be wrong here :)

I do have scripting skills and was intending to do just as you propose, but am failing with setting the right properties using the exiftool. I'll send an example per email.

**Reply by RhetTbull** [reply]
Yeah, I think you're right about not being able to create MakerNotes with exiftool. I tried and got the same result and [exiftool forum](https://exiftool.org/forum/index.php?topic=14268.0) seems to confirm this.  I have another package [CGMetaData](https://github.com/RhetTbull/CGMetadata) that can read/write image metadata using native Apple APIs. It doesn't currently write the AppleMakerNotes but it [might be able to](https://developer.apple.com/documentation/imageio/kcgimagepropertymakerappledictionary?language=objc).  I'll take a look and see what's possible. Would be a useful tool to be able to fix these.

**Reply by Yorian** [reply]
Thanks a lot for the effort. It would be amazing to get this working as well. If I can be of any help: let me know!

**Reply by oPromessa** [reply]
Found this link

https://discussions.apple.com/thread/252183998

Indicating a manual way to convert Google motion photos into Live Photos:

"You can transfer motion photos (android) to iPhone using google photos. Your motion photos must be uploaded to google photos. If you have the same photo on your iPhone you will need to delete the still photo from your library and download from the google photos. It will then play as live/motion photo on either device. This is only successful moving from Samsung Motion Photos to Live Photos."

**Comment by RhetTbull** [comment]
Making some progress. The following prototype python code will write the correct MakerNote to the image file using Core Graphics. I'll need to write separate code using AVAsset to write the metadata to the movie file.

You'll need to `pip install` the following to make this work:

```
    "pyobjc-core>=9.2",
    "pyobjc-framework-Contacts>=9.2",
    "pyobjc-framework-CoreLocation>=9.2",
    "pyobjc-framework-Quartz>=9.2",
    "pyobjc-framework-UniformTypeIdentifiers>=9.2",
    "wurlitzer>=3.0.3",
```

```python
"""Write Maker Note asset id to image file to associate it with a live image"""

from __future__ import annotations

import os
import sys
import uuid
from typing import Any

import objc
import Quartz
from Foundation import (
    NSURL,
    CFDictionaryRef,
    NSData,
    NSMutableData,
    NSMutableDictionary,
)
from wurlitzer import pipes

# what exiftool tool displays as "Content Identifier" is actually a note with value "17"
# this is the value that is used to associate the live image video file
kFigAppleMakerNote_AssetIdentifier = "17"


def image_from_path(
    image_path: str | os.PathLike,
) -> Quartz.CGImageSourceRef:
    """Create a CGImageSourceRef from an image file.

    Args:
        image_path: Path to the image file.

    Returns: CGImageSourceRef
    """
    with objc.autorelease_pool():
        image_url = NSURL.fileURLWithPath_(str(image_path))
        image_source = Quartz.CGImageSourceCreateWithURL(image_url, None)
        if not image_source:
            raise ValueError(f"Could not create image source for {image_path}")
        return image_source


def add_asset_id_to_image(
    image_path: str | os.PathLike,
    asset_id: str,
    destination_path: str | os.PathLike,
) -> None:
    """Write the asset id to file at image_path and save to destination path"""
    image_data = image_from_path(image_path)
    metadata = Quartz.CGImageSourceCopyPropertiesAtIndex(image_data, 0, None)
    metadata_as_mutable = metadata.mutableCopy()
    maker_apple = metadata_as_mutable.objectForKey_(
        Quartz.kCGImagePropertyMakerAppleDictionary
    )
    if not maker_apple:
        maker_apple = NSMutableDictionary.alloc().init()
    maker_apple.setObject_forKey_(asset_id, kFigAppleMakerNote_AssetIdentifier)
    metadata_as_mutable.setObject_forKey_(
        maker_apple, Quartz.kCGImagePropertyMakerAppleDictionary
    )
    write_image_with_metadata(image_data, metadata_as_mutable, destination_path)


def write_image_with_metadata(
    image_data: Quartz.CGImageSourceRef,
    metadata: CFDictionaryRef,
    destination_path: str | os.PathLike,
) -> None:
    """Write image with metadata to destination path"""
    with objc.autorelease_pool():
        image_type = Quartz.CGImageSourceGetType(image_data)
        dest_data = NSMutableData.data()
        destination = Quartz.CGImageDestinationCreateWithData(
            dest_data, image_type, 1, None
        )
        if not destination:
            raise ValueError(
                f"Could not create image destination for {destination_path}"
            )
        with pipes() as (_out, _err):
            # use pipes to catch error messages from CGImageDestinationAddImageFromSource
            # there's a bug in Core Graphics that causes an error similar to
            # AVEBridge Info: AVEEncoder_CreateInstance: Received CreateInstance (from VT)
            # ... AVEBridge Error: AVEEncoder_CreateInstance: returning err = -12908
            # to output to stderr/console but the image is still written correctly
            # reference: https://github.com/biodranik/HEIF/issues/5 and
            # https://forums.developer.apple.com/forums/thread/722204
            Quartz.CGImageDestinationAddImageFromSource(
                destination, image_data, 0, metadata
            )
            Quartz.CGImageDestinationFinalize(destination)
            new_image_data = NSData.dataWithData_(dest_data)
            new_image_data.writeToFile_atomically_(destination_path, True)


def main():
    filename = sys.argv[1]
    asset_id = str(uuid.uuid4()).upper()
    print(f"Adding asset id {asset_id} to {filename}")
    add_asset_id_to_image(filename, asset_id, filename)


if __name__ == "__main__":
    main()

```

**Reply by RhetTbull** [reply]
Doing this for the QuickTime file will be more complex as writing a QuickTime file requires handling multiple audio and video tracks, etc. I'm confident this can be done though.  Then will need to figure out the UI.  I think a command line interface (CLI) with following interface but curious as to what would make it most useful:

```bash
makelive photos/*.jpg photos/*.mov
```

This would find all .jpg + .mov files and find any matching pairs (filepath stem is the same) then assign a new asset ID to both.

But -- Are there cases where you'd want to pair a photo and video whose stem does not match?  Would this be useful and what would the interface be?

Maybe a manual mode:

```bash
makelive --manual PHOTO VIDE
```

It would also be possible to create a native mac app with py2app that provides a service in Finder so you could select two files then right click and select "Make live pair".  I've done this in Python for other projects.  Not sure if this is worth the extra effort though as I'm partial to the command line.

**Comment by RhetTbull** [comment]
@Yorian I've got an initial working script [here](https://github.com/RhetTbull/makelive).  I tested it on the samples you sent and it works (on Ventura 13.5.1). The script is very rudimentary and must be passed the name of the image and the name of the video to be paired. It will overwrite the originals so ensure you have a backup. I've done very little testing.  When I get some more time I'll add a test suite, better UI for the CLI, etc.  Also happy to accept PRs.  I'd like to eventually package this with [PyApp](https://github.com/ofek/pyapp) and get it onto PyPI so it can be pip installed. But I wanted to get a working MVP out quickly. 

I think the script should really scan a list of input files to find matching pairs then apply the tag but currently the photo and video files must be passed as separate arguments.

**Reply by RhetTbull** [reply]
By the way, it worked on the `.mp4` as well as the `.mov`. The `.mp4` was converted to `.mov` upon import.

**Comment by Yorian** [comment]
@RhetTbull You are amazing. Thanks for the quick deep dive you did. Will look into it this evening!

**Reply by RhetTbull** [reply]
One notable issue you should be aware of is that at the moment, XMP metadata stored in the video file is not preserved. I've not found a way to read/write the XMP metadata using native APIs. One potential solution is to use [exiftool](https://exiftool.org/) to export the XMP metadata to an XMP file, process the file, then use exiftool to rewrite the XMP metadata.

- `exiftool -XMP -b -w xmp test.mov`
- `python3 makelive.py test.jpg test.mov`
- `exiftool -tagsfromfile test.xmp -xmp test.mov`

Metadata including XMP and IPTC does appear to be preserved in the image file.

Maybe the command line tool should have a flag for this.

**Reply by RhetTbull** [reply]
It's now on [PyPi](https://pypi.org/project/makelive/) so you just `pipx install makelive`.  Still needs work but this should get you started.

**Reply by Yorian** [reply]
I just tried it with the example file I sent you and it works fine on my side as well! Thank you very much!
